compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/oas/ServiceOasGenerator.java

```diff
@@ -0,0 +1,184 @@
+ * under the License.
+ */
+package io.ballerina.stdlib.http.compiler.oas;
+
+import io.ballerina.compiler.api.SemanticModel;
+import io.ballerina.compiler.api.symbols.Symbol;
+import io.ballerina.compiler.syntax.tree.AnnotationNode;
+import io.ballerina.compiler.syntax.tree.MappingConstructorExpressionNode;
+import io.ballerina.compiler.syntax.tree.MetadataNode;
+import io.ballerina.compiler.syntax.tree.NodeList;
+import io.ballerina.compiler.syntax.tree.ServiceDeclarationNode;
+import io.ballerina.compiler.syntax.tree.SpecificFieldNode;
+import io.ballerina.openapi.service.mapper.ServiceToOpenAPIMapper;
+import io.ballerina.openapi.service.mapper.model.OASGenerationMetaInfo;
+import io.ballerina.openapi.service.mapper.model.OASResult;
+import io.ballerina.openapi.service.mapper.model.ServiceDeclaration;
+import io.ballerina.openapi.service.mapper.model.ServiceNode;
+import io.ballerina.projects.Project;
+import io.ballerina.projects.plugins.AnalysisTask;
+import io.ballerina.projects.plugins.SyntaxNodeAnalysisContext;
+import io.ballerina.tools.diagnostics.DiagnosticSeverity;
+import io.swagger.v3.core.util.Json;
+import io.swagger.v3.oas.models.OpenAPI;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Optional;
+
+import static io.ballerina.openapi.service.mapper.utils.MapperCommonUtils.normalizeTitle;
+import static io.ballerina.stdlib.http.compiler.HttpCompilerPluginUtil.diagnosticContainsErrors;
+import static io.ballerina.stdlib.http.compiler.HttpCompilerPluginUtil.getServiceDeclarationNode;
+
+/**
+ * This class generates the OpenAPI definition resource for the service declaration node.
+ *
+ * @since 2.12.0
+ */
+public class ServiceOasGenerator implements AnalysisTask<SyntaxNodeAnalysisContext> {
+
+    @Override
+    public void perform(SyntaxNodeAnalysisContext context) {
+        if (diagnosticContainsErrors(context)) {
+            return;
+        }
+
+        ServiceDeclarationNode serviceDeclarationNode = getServiceDeclarationNode(context);
+        if (serviceDeclarationNode == null) {
+            return;
+        }
+
+        Optional<AnnotationNode> serviceInfoAnnotation = getServiceInfoAnnotation(serviceDeclarationNode);
+        if (serviceInfoAnnotation.isEmpty()) {
+            return;
+        }
+
+        boolean embedOpenAPI = retrieveValueFromAnnotation(serviceInfoAnnotation.get(), "embed")
+                .map(Boolean::parseBoolean)
+                .orElse(false);
+        if (!embedOpenAPI) {
+            return;
+        }
+
+        Optional<String> contractPath = retrieveValueFromAnnotation(serviceInfoAnnotation.get(), "contract");
+        if (contractPath.isPresent()) {
+            return;
+        }
+
+        Optional<Symbol> symbol = context.semanticModel().symbol(serviceDeclarationNode);
+        if (symbol.isEmpty()) {
+            // Add warning diagnostic
```

Comment:
Is this the missing diagnostic or are we going to add it in separate PR?

Reply:
Ideally it should not go to this check, will remove the comment

Reply:
Removed via [2ac9be7](https://github.com/ballerina-platform/module-ballerina-http/pull/2037/commits/2ac9be7cee3e64184c3452119ece9a10454997b0)